processing/src/main/java/io/druid/segment/DimensionQueryHelper.java

```diff
@@ -0,0 +1,359 @@
+  public void hashValues(Object dimSelector, HyperLogLogCollector collector);
+
+
+  /**
+   * Used by GroupByEngine.
+   *
+   * Return the size, in bytes, of this dimension's values in the grouping key.
+   *
+   * For example, a String implementation would return 4, the size of an int.
+   *
+   * @return size, in bytes, of this dimension's values in the grouping key.
+   */
+  public int getGroupingKeySize();
+
+
+  /**
+   * Used by GroupByEngine.
+   *
+   * A grouping key contains a concatenation of byte[] representations of dimension values.
+   *
+   * When comparing two grouping keys, the individual dimension values will be compared with comparators
+   * provided by the query helper.
+   *
+   * @return A comparator suitable for comparing byte representations of this dimension's type of values.
+   */
+  public Comparator<byte[]> getGroupingKeyByteComparator();
+
+
+  /**
+   * Used by GroupByEngine.
+   *
+   * Perform a relative read on a grouping key ByteBuffer to retrieve a single dimension value, and
+   * add the retrieved value to a GroupBy result map.
+   *
+   * An implementation may choose to not add anything to the result map
+   * (e.g., as the String implementation does for empty rows)
+   *
+   * @param theEvent Result map for the GroupBy query being served
+   * @param outputName The output name of this dimension for the GroupBy query being served, as specified in the DimensionSpec
+   * @param dimSelector Dimension value selector, used for value lookups if needed
+   * @param keyBuffer Grouping key, already positioned at this dimension's offset
+   */
+  public void readDimValueFromGroupingKey(
+      Map<String, Object> theEvent,
+      String outputName,
+      Object dimSelector,
+      ByteBuffer keyBuffer
+  );
+
+
+  /**
+   * Used by GroupByEngine.
+   *
+   * Read the current row from a dimension value selector and add the row values to the grouping key.
+   *
+   * This is called by GroupByEngine's updateValues() function, which uses recursion to traverse the dimensions in the grouping set.
+   *
+   * Before adding a dimension value to the grouping key, this function should duplicate() the provided key buffer and
+   * add the value to the duplicate key.
+   *
+   * After adding a dimension value to the duplicate grouping key, an implementation of this function should call
+   * updateValuesFn on the new key to perform the recursion.
+   *
+   * For multi-value rows, this function should duplicate the original grouping key before adding each value, and
+   * call updateValuesFn on each new key.
+   *
+   * See StringDimensionQueryHelper for a reference implementation.
+   *
+   * @param dimSelector Dimension value selector
+   * @param key ByteBuffer for the grouping key
+   * @param updateValuesFn Function provided by GroupByEngine for updateValues() recursion
+   * @return Return the result of calling updateValuesFn on the updated grouping key
+   */
+  public List<ByteBuffer> addDimValuesToGroupingKey(
```

Comment:
I'm inclined to keep this as-is, see my earlier comment on GroupByEngine v1 deprecation